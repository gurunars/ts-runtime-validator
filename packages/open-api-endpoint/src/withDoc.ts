import { Field } from '@validator/validator'
import { field } from '@validator/validator/registry'
import { Any } from '@validator/validator/util-types'

type Example<T extends Any> = {
  value: T,
  summary: string
}

type Doc<T extends Any> = {
  description?: string,
  format?: string,
  examples?: Record<string, Example<T>>
}

export interface WithDoc<T extends Any, F extends Field<T>> extends Field<T> {
  readonly innerField: F,
  readonly doc: Doc<T>
}

export default field('@validator/fields.WithDoc', <T extends Any, F extends Field<T>> (
  innerField: F,
  doc: Doc<T>
): WithDoc<T, F> => ({
    ...innerField,
    innerField,
    doc
  }))
