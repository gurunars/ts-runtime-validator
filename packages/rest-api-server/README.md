# Rest API server

## The ideas behind

- autocompletion should work out of the box both for the server and for the client
- both server and client should reuse the same logic for:
    - validation
    - serialization
    - type transformation
- meanwhile server defines the pattern matcher, client introspects it

## What should be done for the server?

We actually want to have metatypes for each field
as a part of the interface to reflect the basic structural shape.
The meta types could be used to generate Open API.

## What should be done for the client?

There should be a mean of generating client side library with all the necessary field
types without the actual server side code.

I.e. we generate a client using webpack and tree shaking.

