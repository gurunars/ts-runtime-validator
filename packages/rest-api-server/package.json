{
  "name": "@validator/rest-api-server",
  "license": "Apache-2.0",
  "version": "0.0.1",
  "dependencies": {
    "@validator/validator": "0.0.1",
    "express": "^4.17.1"
  },
  "devDependencies": {
    "@types/express": "^4.17.8"
  }
}
