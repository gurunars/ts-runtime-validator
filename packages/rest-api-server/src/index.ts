export { ServerConfig } from './handler'
export { Route } from './route'
export {
  serve, withMethod,
  GET, HEAD, POST, PUT, DELETE, PATCH
} from './server'
