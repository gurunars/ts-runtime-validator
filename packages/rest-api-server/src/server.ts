import http from 'http'
import { URL } from 'url'
import { StringMapping } from './route'

import { ServerConfig, handle } from './handler-impl'
import { JsonSerialization } from './serialization'
import { Route, RequestSpec, ResponseSpec } from './route'

import { Field } from '@validator/validator'
import { constantField, $, stringField } from '@validator/validator/fields'
import { ConstantField } from '@validator/validator/fields/constantField'

type RequestSpecMethod = Omit<RequestSpec, 'method' | 'pathParams'>

type PathSpec<PathParams extends StringMapping> = typeof $ & Field<PathParams>

// Make it fluid API - to make things work with autocomplete
export const withMethod = <
  Method extends string,
> (method: Method) => <
  PathParams extends StringMapping = StringMapping,
  ReqSpec extends RequestSpecMethod = RequestSpecMethod,
  RespSpec extends ResponseSpec = ResponseSpec
  > (
      pathParams: PathSpec<PathParams>,
      spec: {
        request?: ReqSpec,
        response: RespSpec
      },
      handler: Route<ReqSpec & {
        readonly method: ConstantField<string>,
        readonly pathParams: PathSpec<PathParams>
      }, RespSpec>['handler']
    ): Route => ({
      request: {
        ...(spec.request || {}),
        method: constantField(method),
        pathParams,
      },
      response: spec.response,
      handler: handler as unknown as Route['handler']
    })

export const GET = withMethod('GET')
export const HEAD = withMethod('HEAD')
export const POST = withMethod('POST')
export const PUT = withMethod('PUT')
export const DELETE = withMethod('DELETE')
export const CONNECT = withMethod('CONNECT')
export const OPTIONS = withMethod('OPTIONS')
export const TRACE = withMethod('TRACE')
export const PATCH = withMethod('PATCH')

export const DEFAULT_SERVER_CONFIG: ServerConfig = {
  baseUrl: 'http://localhost:8000',
  serializationFormats: [new JsonSerialization()],
  encoding: 'utf-8',
  frameworkErrorStatusCode: 502,
  appErrorStatusCode: 500,
  reportError: (error: unknown) => {
    console.error(error)
    return Promise.resolve(undefined)
  },
  routes: [
    GET($._('/'),
      {
        response: {
          statusCode: constantField(200),
          data: stringField()
        }
      },
      async () => ({
        statusCode: 200 as const,
        data: 'ROOT'
      })
    )
  ]
}

const mergeServerConfigs = (
  serverConfig: Partial<ServerConfig>
): ServerConfig => ({
  ...DEFAULT_SERVER_CONFIG,
  ...serverConfig,
})

const SUPPORTED_PROTOCOLS= {
  'http': 80,
  'https': 443
} as const

const supportedProtocols = new Set(Object.keys(SUPPORTED_PROTOCOLS))

const verifyProtocol = (protocol: string): protocol is keyof typeof SUPPORTED_PROTOCOLS =>
  supportedProtocols.has(protocol)

const getPort = (baseUrl: string): number => {
  const url = new URL(baseUrl)
  if (!verifyProtocol(url.protocol)) {
    throw `Protocol ${url.protocol} is not supported`
  }
  return url.port ? Number.parseInt(url.port) : SUPPORTED_PROTOCOLS[url.protocol]
}

export const serve = (
  config: Partial<ServerConfig>,
): void => {
  const merged = mergeServerConfigs(config)
  http.createServer(handle.bind(null, merged)).listen(getPort(merged.baseUrl))
}
