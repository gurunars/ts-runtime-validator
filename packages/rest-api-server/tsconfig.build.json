
{
  "extends": "./tsconfig.json",
  "references": [
    { "path": "../validator/tsconfig.build.json" },
  ],
  "exclude": ["src/**/*.spec.ts", "src/**/*.test.ts"]
}
