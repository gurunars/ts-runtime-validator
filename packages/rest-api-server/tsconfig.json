{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "composite": true,
    "paths": {
      "@validator/validator/*": ["../validator/src/*"],
      "@validator/validator": ["../validator/src/index.ts"],
      "@validator/rest-api-server/*": ["../rest-api-server/src/*"],
      "@validator/rest-api-server": ["../rest-api-server/src/index.ts"],
      "@validator/open-api-endpoint/*": ["../open-api-endpoint/src/*"],
      "@validator/open-api-endpoint": ["../open-api-endpoint/src/index.ts"]
    }
  },
  "references": [
    { "path": "../validator" }
  ]
}
