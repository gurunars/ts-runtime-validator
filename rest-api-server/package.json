{
  "name": "@validator/rest-api-server",
  "license": "Apache-2.0",
  "version": "0.0.1",
  "dependencies": {
    "@validator/validator": "0.0.1"
  },
  "scripts": {
    "build": "ts-node ../build.ts"
  }
}
