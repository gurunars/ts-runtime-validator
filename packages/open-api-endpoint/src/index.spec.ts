import { GET, POST, Route } from '@validator/rest-api-server'
import {
  $, stringField, objectField, constantField, booleanField, numberField, arrayField
} from '@validator/validator/fields'
import genOpenApi from '.'
import { DEFAULT_SERVER_CONFIG } from '@validator/rest-api-server/server'
import withDoc from './withDoc'

test('fullRoute', () => {
  const routes: Route[] = [
    POST($._('/items'),
      {
        request: {
          data: objectField({
            title: stringField(/.*/),
            count: numberField()
          })
        },
        response: {
          statusCode: constantField(201)
        }
      },
      async () => Promise.resolve({
        statusCode: 201 as const,
      })
    ),
    GET($._('/item/')._('id', stringField()),
      {
        request: {
          headers: objectField({
            key: withDoc(numberField(), {
              description: 'key header',
              examples: {
                sampleKey: {
                  value: 13 as number,
                  summary: 'Sample value'
                }
              }
            })
          }),
          queryParams: objectField({
            flag: booleanField(),
          })
        },
        response: {
          statusCode: constantField(200),
          data: objectField({
            items: arrayField(numberField())
          })
        },
      },
      async () => Promise.resolve({
        statusCode: 200 as const,
        data: {
          items: [1, 2, 3]
        }
      })
    )
  ]

  expect(genOpenApi({
    ...DEFAULT_SERVER_CONFIG,
    info: {
      title: 'Test',
      version: '1.0.0'
    }
  }, routes)).toMatchSnapshot()


})
