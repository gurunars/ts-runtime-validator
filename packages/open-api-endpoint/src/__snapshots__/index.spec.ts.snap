// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fullRoute 1`] = `
Object {
  "info": Object {
    "title": "Test",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/item/{id}": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "query",
            "name": "flag",
            "required": true,
            "schema": Object {
              "type": "boolean",
            },
          },
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "items": Object {
                      "items": Object {
                        "type": "integer",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "items",
                  ],
                },
              },
            },
            "description": "",
          },
        },
      },
    },
    "/items": Object {
      "post": Object {
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "properties": Object {
                  "count": Object {
                    "type": "integer",
                  },
                  "title": Object {
                    "pattern": ".*",
                    "type": "string",
                  },
                },
                "required": Array [
                  "title",
                  "count",
                ],
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "201": Object {
            "description": "",
          },
        },
      },
    },
  },
  "servers": Array [
    Object {
      "url": "http://localhost:8000",
    },
  ],
}
`;
